perfkitbenchmarker/scripts/object_storage_api_tests.py

```diff
@@ -518,6 +643,240 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+    num_alive = sum((thread.isAlive() for thread in threads))
+    if num_alive == 0:
+      break
+    else:
+      logging.info('%s of %s threads are still working.',
+                   num_alive, FLAGS.num_streams)
+  logging.info('All threads complete.')
+
+  results = []
+
+  try:
+    while True:
+      res = result_queue.get_nowait()
+      results.append(res)
+  except Queue.Empty:
+    pass
+
+  return results
+
+
+def MultiStreamWrites(storage_schema, host_to_connect=None):
+  """Run multi-stream write threads."""
+
+  size_distribution = yaml.load(cStringIO.StringIO(FLAGS.object_sizes))
+
+  payload_bytes, payload_string = GenerateWritePayload(
+      MaxSizeInDistribution(size_distribution))
+
+  results = RunThreadedWorkers(
+      WriteWorker,
+      (storage_schema,
+       host_to_connect,
+       payload_bytes,
+       payload_string,
+       size_distribution,
+       FLAGS.objects_per_stream,
+       FLAGS.start_time))
+
+  # object_records is the data we leave on the VM for future reads. We
+  # need to pass data to the reader so it will know what object names
+  # it can read. The file format is a JSON list of lists in the format
+  #  [ [object_name_1, object_size_1],
+  #    [object_name_2, object_size_2],
+  #    ... ]
+  if FLAGS.objects_written_file is not None:
+    object_records = []
+    for result in results:
+      for name, size in zip(result['object_names'], result['sizes']):
+        object_records.append([name, size])
+    if os.path.exists(FLAGS.objects_written_file):
+      os.remove(FLAGS.objects_written_file)
+    with open(FLAGS.objects_written_file, 'w') as out:
+      json.dump(object_records, out)
+
+  # operation_records is the data we send back to the controller.
+  operation_records = []
+  for result in results:
+    operation = result['operation']
+    stream_num = result['stream_num']
+    for start_time, latency, size in zip(result['start_times'],
+                                         result['latencies'],
+                                         result['sizes']):
+      operation_records.append({'operation': operation,
+                                'start_time': start_time,
+                                'latency': latency,
+                                'size': size,
+                                'stream_num': stream_num})
+  json.dump(operation_records, sys.stdout, indent=0)
+
+
+def MultiStreamReads(storage_schema, host_to_connect=None):
+  """Run multi-stream read threads."""
+
+  # Read the object records that the MultiStreamWriter left for us.
+  if FLAGS.objects_written_file is None:
+    raise ValueError('The MultiStreamRead benchmark needs a list of object '
+                     'names to read from. Use '
+                     '--objects_written_file=<filename>.')
+  with open(FLAGS.objects_written_file, 'r') as object_file:
+    object_records = json.load(object_file)
+
+  num_workers = FLAGS.num_streams
+  objects_by_worker = []
```

Comment:
please extract this logic out into a function, and write unit-test for it. 

Please decide if you plan to support the case where num-workers > object_records, if so, then current logic would fail. (based on my cursory simulation run in my head for the case of object_records = 3 and num_workers=4), you will run into array out of bounds exception.